nemo/collections/nlp/models/dialogue_state_tracking_generative/dialogue_gpt_model.py

```diff
@@ -146,23 +235,58 @@ def forward(self, input_ids, attention_mask, labels):
         if self.cfg.library == "huggingface":
             output = self.language_model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)
             loss = output['loss']
+
         elif self.cfg.library == "megatron":
-            position_ids = torch.arange(input_ids.size(1), dtype=torch.long, device=input_ids.device)
+            num_prompt_tokens = (
+                self.language_model.num_prompt_tokens if hasattr(self.language_model, 'num_prompt_tokens') else 0
+            )
+            position_ids = torch.arange(
+                start=num_prompt_tokens,
+                end=num_prompt_tokens + input_ids.size(1),
+                dtype=torch.long,
+                device=input_ids.device,
+            )
+
             position_ids = position_ids.unsqueeze(0).repeat(input_ids.size(0), 1)
+
+            # 'assit_intent_and_slot' has prompt_id of 1
+            # 'assit_intent_and_slot_with_options' has prompt_id of 2
+            prompt_ids = torch.tensor([1] * input_ids.size(0)) if self.prompt_tags else None
+            if self.cfg.dataset.prompt_template == "prompt_tuning_with_options" and prompt_ids is not None:
+                prompt_ids = prompt_ids * 2
+            attn_mask_add_on = torch.ones((attention_mask.size(0), num_prompt_tokens), device=attention_mask.device)
+            full_attention_mask = torch.cat([attn_mask_add_on, attention_mask], axis=-1)
+            full_attention_mask_expand = torch.tril(
+                full_attention_mask.unsqueeze(2).tile(full_attention_mask.size(1))
+            ).unsqueeze(1)
+
+            attn_mask = full_attention_mask_expand > 0
+
+            prompt_token_labels = torch.full(
+                size=(input_ids.size(0), num_prompt_tokens),
+                fill_value=self.tokenizer.tokenizer.pad_token_id,
+                dtype=torch.long,
+                device=input_ids.device,
+            )
+
+            input_ids_new = torch.cat([prompt_token_labels, input_ids], axis=1)
```

Comment:
If I don't append the inputs_id and the attention mask, it raises the error `File "/home/zhilinw/NeMo/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py", line 220, in forward
    output_tensor = self.model(tokens, text_position_ids, attention_mask, labels=labels, prompt_ids=prompt_ids,)
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/zhilinw/NeMo/nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py", line 178, in forward
    lm_output = self.language_model(
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/zhilinw/NeMo/nemo/collections/nlp/modules/common/megatron/language_model.py", line 743, in forward
    encoder_output = self.encoder(
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/zhilinw/NeMo/nemo/collections/nlp/modules/common/megatron/transformer.py", line 855, in forward
    hidden_states = layer(
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/zhilinw/NeMo/nemo/collections/nlp/modules/common/megatron/transformer.py", line 633, in forward
    return super().forward(
  File "/home/zhilinw/NeMo/nemo/collections/nlp/modules/common/megatron/transformer.py", line 533, in forward
    attention_output, attention_bias = self.self_attention(
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/zhilinw/NeMo/nemo/collections/nlp/modules/common/megatron/transformer.py", line 356, in forward
    attention_probs = self.scale_mask_softmax(attention_scores, attention_mask)
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/apex/transformer/functional/fused_softmax.py", line 149, in forward
    return self.forward_torch_softmax(input, mask)
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/apex/transformer/functional/fused_softmax.py", line 184, in forward_torch_softmax
    mask_output = self.mask_func(input, mask) if mask is not None else input
  File "/home/zhilinw/NeMo/nemo/collections/nlp/modules/common/megatron/utils.py", line 82, in attention_mask_func
    attention_scores.masked_fill_(attention_mask, -10000.0)
RuntimeError: The expanded size of the tensor (278) must match the existing size (128) at non-singleton dimension 3.  Target sizes: [8, 12, 278, 278].  Tensor sizes: [8, 1, 128, 128]`

Reply:
Otherwise, if I don't append the input_ids only, it raises the error `  File "/home/zhilinw/NeMo/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py", line 220, in forward
    output_tensor = self.model(tokens, text_position_ids, attention_mask, labels=labels, prompt_ids=prompt_ids,)
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/zhilinw/NeMo/nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py", line 189, in forward
    return post_language_model_processing(
  File "/home/zhilinw/NeMo/nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py", line 67, in post_language_model_processing
    loss = tensor_parallel.vocab_parallel_cross_entropy(output.float(), labels)
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/apex/transformer/tensor_parallel/cross_entropy.py", line 103, in vocab_parallel_cross_entropy
    return _VocabParallelCrossEntropy.apply(vocab_parallel_logits, target)
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/apex/transformer/tensor_parallel/cross_entropy.py", line 53, in forward
    predicted_logits_1d = logits_2d[arange_1d, masked_target_1d]
IndexError: shape mismatch: indexing tensors could not be broadcast together with shapes [2224], [1024]`

Reply:
if I don't append the input_ids, it raises the error `  File "/home/zhilinw/NeMo/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py", line 220, in forward
    output_tensor = self.model(tokens, text_position_ids, attention_mask, labels=labels, prompt_ids=prompt_ids,)
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/zhilinw/NeMo/nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py", line 189, in forward
    return post_language_model_processing(
  File "/home/zhilinw/NeMo/nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py", line 67, in post_language_model_processing
    loss = tensor_parallel.vocab_parallel_cross_entropy(output.float(), labels)
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/apex/transformer/tensor_parallel/cross_entropy.py", line 103, in vocab_parallel_cross_entropy
    return _VocabParallelCrossEntropy.apply(vocab_parallel_logits, target)
  File "/home/zhilinw/anaconda3/lib/python3.9/site-packages/apex/transformer/tensor_parallel/cross_entropy.py", line 53, in forward
    predicted_logits_1d = logits_2d[arange_1d, masked_target_1d]
IndexError: shape mismatch: indexing tensors could not be broadcast together with shapes [2224], [1024]`

Reply:
Might you know where the bug might lie?

Reply:
Also, if this implementation is incorrect, it's probably quite unlikely for 1.3B model to achieve intent acc of 60+ right?